src/include/migraphx/op/concat.hpp

```diff
@@ -73,49 +74,103 @@ struct concat
         }
         return offsets;
     }
+
     shape normalize_compute_shape(std::vector<shape> inputs) const
     {
+        // Todo:: not using check_shapes struct; it
+        // doesn't have a method to check for at least one shape.
+        // A single input is a no-op, but it is legal
         if(inputs.empty())
         {
-            MIGRAPHX_THROW("CONCAT: Number of input tensors should exceed 0");
+            MIGRAPHX_THROW("CONCAT: called with empty input list");
         }
 
-        const auto& first_shape_lens = inputs.front().lens();
-        const auto& type             = inputs.front().type();
-        for(std::size_t l = 0; l < first_shape_lens.size(); l++)
+        if(not std::all_of(inputs.begin(), inputs.end(), [&](const shape& s) {
+               return (s.ndim() == inputs[0].ndim() and s.type() == inputs[0].type());
+           }))
+        {
+            MIGRAPHX_THROW("CONCAT: input tensors are not all the same rank or type");
+        }
+
+        if(axis >= inputs[0].ndim())
+            MIGRAPHX_THROW("CONCAT: axis attribute out of range");
+
+        if(std::none_of(inputs.begin(), inputs.end(), [&](const shape& s) { return s.dynamic(); }))
         {
-            if(l != axis)
+            // Static input shapes
+            const auto& first_shape_lens = inputs.front().lens();
+            const auto& type             = inputs.front().type();
+            for(std::size_t l = 0; l < first_shape_lens.size(); l++)
             {
-                if(not std::all_of(inputs.begin(), inputs.end(), [&](auto s) {
-                       return s.lens()[l] == first_shape_lens[l];
-                   }))
+                if(l != axis)
                 {
-                    MIGRAPHX_THROW("CONCAT: Non-axis dimensions should match");
+                    if(not std::all_of(inputs.begin(), inputs.end(), [&](auto s) {
+                           return s.lens()[l] == first_shape_lens[l];
+                       }))
+                    {
+                        MIGRAPHX_THROW("CONCAT: all input dimensions should match along axis " +
+                                       std::to_string(l));
+                    }
                 }
             }
+            std::size_t new_dim_axis = 0;
+            for(const auto& input : inputs)
+            {
+                const auto& lens = input.lens();
+                new_dim_axis += lens[axis];
+            }
+            std::vector<std::size_t> new_lens;
```

Comment:
You can initialize new_lens directly with `first_shape_lens` instead of first declaring and then copying.

Reply:
done